ImputationPipeline/CreateAggregationSets/CreateAggregationSets.py

```diff
@@ -0,0 +1,235 @@
+        lab_batch_sample_sets_dict = dict()  # dict from lab_batch to highest group aggregation_set for that lab_batch
+
+        if f'{table_name}_set' in self.available_tables:
+            # Download current sample_set table
+            print(f'Downloading {table_name}_set table...')
+            sample_set_response = fapi.get_entities(self.workspace_namespace, self.workspace_name, f'{table_name}_set')
+            if not sample_set_response.ok:
+                raise RuntimeError(f'ERROR: {sample_set_response.text}')
+            sample_sets_dict = json.loads(sample_set_response.text)
+            existing_sample_sets = {s['name']: [e['entityName'] for e in s['attributes'][f'{table_name}s']['items']] for
+                                    s
+                                    in sample_sets_dict}
+
+            for sample_set in sample_sets_dict:
+                samples = [e['entityName'] for e in sample_set['attributes'][f'{table_name}s']['items']]
+                samples_already_in_aggregation_sets.update(samples)
+
+            for sample_set in sample_sets_dict:
+                attributes = sample_set['attributes']
+                lab_batch = attributes['lab_batch']
+                this_aggregation_set = pre_existing_aggregation_set(lab_batch, attributes['group'],
+                                                                    attributes['delivered'])
+                if lab_batch in lab_batch_sample_sets_dict:
+                    if lab_batch_sample_sets_dict[lab_batch].group < this_aggregation_set.group:
+                        if not lab_batch_sample_sets_dict[lab_batch].delivered:
+                            raise RuntimeError(
+                                f'Aggregation set {lab_batch_sample_sets_dict[lab_batch].set_id} has not been delivered, '
+                                f'but later set {this_aggregation_set.set_id} also exists')
+                        lab_batch_sample_sets_dict[attributes['lab_batch']] = this_aggregation_set
+                else:
+                    lab_batch_sample_sets_dict[attributes['lab_batch']] = this_aggregation_set
+
+        # Read samples from samples table
+        print(f'Reading {table_name} table...')
+        sample_response = fapi.get_entities(self.workspace_namespace, self.workspace_name, f'{table_name}')
+        if not sample_response.ok:
+            raise RuntimeError(f'ERROR: {sample_response.text}')
+
+        samples = json.loads(sample_response.text)
+        # Writing new sample_set_membership.tsv
+        added_sample_sets_dict = dict()  # dictionary from lab_batch to aggregation sets with added samples
+        control_samples_dict = dict()  # dictionary from lab_batch to sample id of control sample
+        added_samples_dict = dict()  # dictionary from set_id to list of samples to be added to the set
+        with open(f'CreateSampleSets_data/new_{table_name}_set_membership.tsv', 'w') as new_membership_file:
```

Comment:
Looking at [this line](https://github.com/broadinstitute/fiss/blob/master/firecloud/api.py#L272) in the firecloud api it seems like `upload_entities_tsv` supports the `entities_tsv` parameter to be a `StringIO`. If you use that you'd be independent of the file system and wouldn't have to create and delete directories

Reply:
oh cool, that's way nicer